### DISCUSSION TITLE
Slow framerate on some software that reads OBS Virtual Camera

### DISCUSSION THREAD

**Discussion by janmonterrubio** [letmaik/pyvirtualcam#94] [Q&A] [answered]
I'm using a copy of the web_cam filter program to just use the camera as a passthrough for some testing (will add open cv on top of it).

Here's the modified version:

```python
# This script is the sample from the README.

import cv2

import pyvirtualcam

cap = cv2.VideoCapture(1)

pref_width = 1280
pref_height = 720
pref_fps_in = 30
cap.set(cv2.CAP_PROP_FRAME_WIDTH, pref_width)
cap.set(cv2.CAP_PROP_FRAME_HEIGHT, pref_height)
cap.set(cv2.CAP_PROP_FPS, pref_fps_in)

# native format UYVY
with pyvirtualcam.Camera(width=pref_width, height=pref_height, fps=15, fmt=pyvirtualcam.PixelFormat.BGR, print_fps=True) as cam:
    print(f'Using virtual camera: {cam.device}')
    print(f'Using {cam.native_fmt}')
     # RGB
    while True:
        ret, frame = cap.read()

        og_frame = frame

        cam.send(og_frame)
        cam.sleep_until_next_frame()
        cv2.imshow('preview', og_frame)
        cv2.waitKey(1)

```

I noticed there's slow (really slow about 5ps) frame rate on MS Teams when it reads from the `pyvirtualcamera`:

![teams_pycam_trim](https://user-images.githubusercontent.com/102086917/186966720-ac2b9342-3d6a-4e5e-8dd7-de4fc3e0b876.gif)

Compared to just using obs directly:

![teams_obs_trim](https://user-images.githubusercontent.com/102086917/186966798-27947131-03b8-4745-8403-e9f947b6bc31.gif)

For sanity, I also compared zoom with pyvirtualcamera and obs and the frame rate didn't seem to differ:

![zoom_obs_trim](https://user-images.githubusercontent.com/102086917/186966894-1de646a7-336a-433c-996f-640e2fee9892.gif)
![zoom_pycam_trim](https://user-images.githubusercontent.com/102086917/186966939-91fa4b4e-b0ac-4640-886f-1de8aa7eaaf2.gif)

I'm not entirely sure what other things I can do to diagnose this.

I compared a bit what exists in: https://github.com/letmaik/pyvirtualcam/tree/main/pyvirtualcam/native_macos_obs vs https://github.com/obsproject/obs-studio/tree/master/plugins/mac-virtualcam

And there seems to be a couple of changes around a memory leak and some IOSource stuff.

I'm on an M1 Mac. 

To further muddle this, a test on a non M1 mac doesn't seem to suffer from the same problem, so I'm guessing it may be related to something with newer macs. 

I appreciate any insight you have and any time you take to read this through.

## Specs

* Mac OS Monterrey 12.5
* Apple M1 Pro
* Memory 16GB
* OBS 27.2.4
* Pyvirtualcam 0.9.1 (built from source)

**Comment by letmaik** [comment]
On the non M1 Mac, did that run the same macOS version and OBS version?

**Reply by willpruyn** [reply]
👋 hey! I was the one running on the non M1 mac. I'm running Catalina 10.15.1 for my maxOS version and 27.2.4 (64 bit) for my OBS version. Hope it helps!

**Reply by letmaik** [reply]
Odd, I'm not sure what the issue could be. OBS 28 will have some changes to avoid unnecessary copies of frames when using the Mac virtual camera, but I doubt it'll have any effect on this issue.
Can you try other video apps as well? It seems odd that it is just Teams. Also, in Zoom, does enabling the HD option make any difference?

**Reply by janmonterrubio** [reply]
I think you may be onto something @letmaik , I tried this in slack and it also has the same behavior. 

I'm wondering if this is a downstream electron problem... since zoom is non electron based but slack and msteams are afaik. 

I'm going to try to spring up a small electron app that just reads from this device and report back on the differences. 

I'll check the zoom option too.

**Reply by janmonterrubio** [reply]
While i get on the electron app, does this output format give you any ideas if it could be the difference?

<img width="792" alt="image" src="https://user-images.githubusercontent.com/102086917/188648787-a1a41cdf-113a-4356-8157-e327a7197515.png">


I believe the cam nativ_fmt was UVYV when I was printing that out.

**Reply by janmonterrubio** [reply]
![obs_electron](https://user-images.githubusercontent.com/102086917/188655281-a1983588-7930-43a7-b4d2-e35ec3d47533.gif)
![electron_pycam](https://user-images.githubusercontent.com/102086917/188655311-6a114741-c677-4671-a1d2-450d7bf41779.gif)


Here's the samples for a simple electron App. I think this is probably something with the video format and electron? that's the one consistent difference I can notice.

**Reply by letmaik** [reply]
Very odd, thanks for replicating this.
The native format of the mac OBS virtual cam is UYVY. pyvirtualcam converts the input frames that it gets from Python (in your script BGR) automatically into UYVY and that happens in C++ and is fast. It happens in all cases, both Zoom and Teams etc., so that can't be the issue.
Do you see the same issue in Safari? For example, using the online version of Teams or some webcam test page.

**Reply by janmonterrubio** [reply]
Unfortunately, safari does not recognize OBS virtual cam.

However I did download firefox and tried it.

![firefox_native](https://user-images.githubusercontent.com/102086917/188977993-537fb746-af96-471a-8763-36491870c958.gif)

![firefox_pyvirtual](https://user-images.githubusercontent.com/102086917/188977603-8c3d2e32-40e3-48b1-adb0-c6c5a2b49464.gif)


_so this may be a chromium bug_ :(

**Reply by IDDD-Dennis** [reply]
Facing similar issue where virtual camera works with Zoom on Mac M1 but not Teams or Slack. Were you able to find a solution?

**Reply by janmonterrubio** [reply]
Negative, but I also have not tried the latest version of OBS. Are you using the updated 28.x OBS and pyvirtualcam 0.10ish (you probably have to build from source), that's what I had to do.

**Comment by janmonterrubio** [comment, marked as answer]
This can be closed, see #100
